scripts/lib/CIME/XML/generic_xml.py

```diff
@@ -48,18 +52,114 @@ def read(self, infile, schema=None):
         logger.debug("read: " + infile)
         with open(infile, 'r') as fd:
             if self.tree:
-                self.root.append(ET.parse(fd).getroot())
+                self.add_child(__Element(ET.parse(fd).getroot()))
             else:
                 self.tree = ET.parse(fd)
-                self.root = self.tree.getroot()
+                self.root = __Element(self.tree.getroot())
 
         if schema is not None and self.get_version() > 1.0:
             self.validate_xml_file(infile, schema)
 
         logger.debug("File version is {}".format(str(self.get_version())))
 
+    #
+    # API for individual node operations
+    #
+
+    def get(self, node, attrib_name):
+        return node.xml_element.get(attrib_name)
+
+    def set(self, node, attrib_name, value):
+        return node.xml_element.set(attrib_name, value)
+
+    def has(self, node, attrib_name):
+        return attrib_name in node.xml_element.attrib
+
+    def set_name(self, node, name):
+        node.xml_element.tag = name
+
+    def set_text(self, node, text):
+        node.xml_element.text = text
+
+    def name(self, node):
+        return node.xml_element.tag
+
+    def text(self, node):
+        return node.xml_element.text
+
+    def add_child(self, node, root=None):
+        """
+        Add element node to self at root
+        """
+        root = root if root else self.root
+        root.append(node)
+
+    def make_child(self, name, text=None, root=None, attributes=None):
+        root = root if root else self.root
+        node = __Element(ET.SubElement(root, name, attrib=attributes))
+        if text:
+            self.set_text(node, text)
+
+        return node
+
+    def get_children(self, root=None, name=None, attributes=None):
```

Comment:
I think a key to better performance will be to primary limit searches to direct children.